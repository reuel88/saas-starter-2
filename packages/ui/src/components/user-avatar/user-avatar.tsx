import { cn } from "@repo/utilities";
import { cva, VariantProps } from "class-variance-authority";
import { FC } from "react";
import { Avatar, AvatarFallback, AvatarImage } from "../ui/avatar";

export const avatarVariants = cva("", {
  variants: {
    size: {
      xs: "h-6 w-6",
      sm: "h-8 w-8",
      md: "h-10 w-10",
      lg: "h-12 w-12",
      xl: "h-14 w-14",
    },
  },
  defaultVariants: {
    size: "md",
  },
});

type UserAvatarProps = VariantProps<typeof avatarVariants> & {
  className?: string;
  firstName?: string;
  lastName?: string;
  name?: string;
  imageSrc?: string;
};

export const UserAvatar: FC<UserAvatarProps> = ({
  className = "",
  firstName,
  lastName,
  name,
  imageSrc,
  size,
}) => {
  const haveFullName = firstName && lastName;
  const avatarName = haveFullName ? `${firstName} ${lastName}` : name || "";

  const moniker = haveFullName
    ? `${firstName.substring(0, 1)}${lastName.substring(0, 1)}`
    : avatarName.substring(0, 1);

  return (
    <Avatar className={cn(avatarVariants({ size, className }))}>
      <AvatarImage src={imageSrc} alt={avatarName} />
      <AvatarFallback>{moniker.toUpperCase()}</AvatarFallback>
    </Avatar>
  );
};
