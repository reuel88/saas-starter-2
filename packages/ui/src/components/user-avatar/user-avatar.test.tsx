import { render, screen } from "@testing-library/react";
import { UserAvatar } from "./user-avatar";

describe("UserAvatar", () => {
  const originalGlobalImage = window.Image;

  beforeEach(() => {
    (window.Image as any) = class MockImage {
      onload: () => void = () => {};
      src: string = "";
      constructor() {
        setTimeout(() => {
          this.onload();
        }, 300);
        return this;
      }
    };
  });

  afterEach(() => {
    window.Image = originalGlobalImage;
  });

  it("should render the fallback 'JS' initially", () => {
    // Arrange

    // Act
    const { baseElement } = render(
      <UserAvatar firstName="John" lastName="Smith" />,
    );
    const fallback = screen.getByText("JS");

    // Assert
    expect(fallback).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render the fallback 'J' initially", () => {
    // Arrange

    // Act
    const { baseElement } = render(<UserAvatar name="John Smith" />);
    const fallback = screen.getByText("J");

    // Assert
    expect(fallback).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render the fallback empty string initially", () => {
    // Arrange

    // Act
    const { baseElement } = render(<UserAvatar />);
    const jsfallback = screen.queryByText("JS");
    const jfallback = screen.queryByText("J");

    // Assert
    expect(jsfallback).not.toBeInTheDocument();
    expect(jfallback).not.toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render an AvatarImage component with the provided image source", async () => {
    // Arrange
    const imageSrc = "https://github.com/shadcn.png";

    // Act
    const { baseElement } = render(
      <UserAvatar imageSrc={imageSrc} firstName="John" lastName="Smith" />,
    );
    const avatarImage = await screen.findByRole("img");

    // Assert
    expect(avatarImage).toHaveAttribute("src", imageSrc);
    expect(baseElement).toMatchSnapshot();
  });
});
