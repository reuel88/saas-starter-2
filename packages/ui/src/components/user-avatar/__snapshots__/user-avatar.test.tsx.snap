// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserAvatar > should render an AvatarImage component with the provided image source 1`] = `
<body>
  <div>
    <span
      class="relative flex shrink-0 overflow-hidden rounded-full h-10 w-10"
    >
      <img
        alt="John Smith"
        class="aspect-square h-full w-full"
        src="https://github.com/shadcn.png"
      />
    </span>
  </div>
</body>
`;

exports[`UserAvatar > should render the fallback 'J' initially 1`] = `
<body>
  <div>
    <span
      class="relative flex shrink-0 overflow-hidden rounded-full h-10 w-10"
    >
      <span
        class="flex h-full w-full items-center justify-center rounded-full bg-muted"
      >
        J
      </span>
    </span>
  </div>
</body>
`;

exports[`UserAvatar > should render the fallback 'JS' initially 1`] = `
<body>
  <div>
    <span
      class="relative flex shrink-0 overflow-hidden rounded-full h-10 w-10"
    >
      <span
        class="flex h-full w-full items-center justify-center rounded-full bg-muted"
      >
        JS
      </span>
    </span>
  </div>
</body>
`;

exports[`UserAvatar > should render the fallback empty string initially 1`] = `
<body>
  <div>
    <span
      class="relative flex shrink-0 overflow-hidden rounded-full h-10 w-10"
    >
      <span
        class="flex h-full w-full items-center justify-center rounded-full bg-muted"
      />
    </span>
  </div>
</body>
`;
